doc/developer_guide.md

```diff
@@ -0,0 +1,418 @@
+function upon receiving the byte payload. To achieve better performance, both
+the libfabric library and ucx library provides APIs to directly send messages
+with tag matching and RDMA operations that avoid extra data copy and shorten
+latencies. Thus both netmod have direct implementations of netmod API that
+skips active message.
+
+Both netmod provide direct RMA operations where it can, and fallback to active
+message where there is limitations. All window synchronizations are based on
+active messages.
+
+## Collectives
+
+Collectives are implemented in `src/mpi/coll/`, with separate folders for each
+collective function.
+
+Use `MPI_Bcast` for example, `bcast/bcast.c` defines 3 functions, mostly
+boilerplates. The binding layer will call **`MPIR_Bcast`**, which will call
+`MPID_Bcast` unless the control variables (CVAR) directs differently.
+Typically `MPID_Bcast` will fallback to call **`MPIR_Bcast_impl`**,  which
+will check CVARs and select particlar algorithm.  By default, it will call
+**`MPIR_Bcast_allcomm_auto`**, which will automatically choose a best
+algorithm based on selection logic defined by a runtime json file.
+
+Each algorithm is typically implemented in a separate file.
+
+## Romio
+
+MPI-IO (MPI_File) APIs are implemented as Romio library. Romio is implemented
+entirely on top of other MPI functions and doesn't use MPICH internals. This
+is similar in a way as how we implement the Fortran bindings.
+
+## MPI_T
+
+MPI Tools information interface provides orthogonal system for controlling and
+monitoring implementation behavior by means of control variables (CVAR),
+performance variables (PVAR), and the event callback interface.
+
+## MPL
+
+All utilities, whose functionality is indepeendent of MPICH internals, are
+implemented inside MPL. All functions from MPL uses `MPL_` prefix.
+
+Some of the essential MPICH functionality are provided by MPL. They include:
+* Debug logging
+* Atomic variables
+* Thread functions
+* Memory tracing
+* GPU support
+
+Sometime we use an additional MPIR wrapping layer on top of MPL functions to
+provide additional control. For example, the threading interfaces are wrapped
+in `src/mpid/common/thread/mpidu_thread_fallback.h` to allow device override
+of locking mechanism. Another example, GPU interface are wrapped to allow
+`MPIR_CVAR_ENABLE_GPU` to by pass GPU functions (when it is set to disable).
+
+## CVAR
+
+Control variables are a mechanism to control library behavior using
+environment variables. The MPI Tool interface allows usage of control
+variables in code during runtime as well (however, it is rather clunky
+thus not widely used).
+
+Reference:
+* [notes on CVAR](notes/cvar.md)
+
+## PM and PMI
+References:
+* [Notes on Process Manager](notes/pm.md)
+* [Notes on Process Manager Interface](notes/pmi.md)
+* [Notes on Namepub](notes/namepub.md)
+* [Using the Hydra Process Manager](https://wiki.mpich.org/mpich/index.php/Using_the_Hydra_Process_Manager)
```

Comment:
This should link to the new wiki page,
https://github.com/pmodels/mpich/blob/main/doc/wiki/how_to/Using_the_Hydra_Process_Manager.md